File: hard/0004_median_of_two_sorted_arrays/solution.cpp
#include <vector>

class Solution {
public:
  double findMedianSortedArrays(std::vector<int> &nums1,
                                std::vector<int> &nums2) {
    std::vector<int>::iterator it1 = nums1.begin();
    std::vector<int>::iterator it2 = nums2.begin();

    std::vector<int>::iterator prev, curr;

    for (size_t _ = 0; _ < (nums1.size() + nums2.size()) / 2 + 1; ++_) {
      prev = curr;

      if (it2 == nums2.end() || (it1 != nums1.end() && *it1 < *it2))
        curr = it1++;
      else
        curr = it2++;
    }

    if ((nums1.size() + nums2.size()) % 2 == 0)
      return (float)(*prev + *curr) / 2;
    else
      return *curr;
  }
};

File: hard/0004_median_of_two_sorted_arrays/solution.py
from typing import List


class Solution:
    def findMedianSortedArrays(
        self, nums1: List[int], nums2: List[int]
    ) -> float:
        nums1_len = len(nums1)
        nums2_len = len(nums2)
        nums_len = nums1_len + nums2_len

        i1 = i2 = 0
        curr = None

        for _ in range(nums_len // 2 + 1):
            prev = curr

            if i2 == nums2_len or (i1 != nums1_len and nums1[i1] < nums2[i2]):
                curr = nums1[i1]
                i1 += 1
            else:
                curr = nums2[i2]
                i2 += 1

        if nums_len % 2 == 0:
            return (prev + curr) / 2
        else:
            return curr
